tiled/_tests/test_authentication.py

```diff
@@ -449,3 +449,151 @@ def test_sticky_identity(enter_password, config):
+        context.http_client.get(url_with_key).raise_for_status()
+        context.http_client.cookies = None
+        # The same endpoint fails without an API key
+        with fail_with_status_code(401):
+            context.http_client.get(resource).raise_for_status()
+
+
+def test_admin_api_key_any_principal_exceeds_scopes(enter_password, principals_context):
+    """
+    Admin cannot create API key that exceeds scopes for another principal.
+    """
+    with principals_context["context"] as context:
+        # Log in as Alice, create and use API key after logout
+        with enter_password("secret1"):
+            context.authenticate(username="alice")
+
+        principal_uuid = principals_context["uuid"]["bob"]
+        with fail_with_status_code(400) as fail_info:
+            _create_api_key_other_principal(
+                context=context, uuid=principal_uuid, scopes=["read:principals"]
+            )
+            fail_message = " must be a subset of the principal's scopes "
+            assert fail_message in fail_info.response.text
+        context.logout()
+
+
+@pytest.mark.parametrize("username", ("alice", "bob"))
+def test_api_key_any_principal(enter_password, principals_context, username):
+    """
+    Ordinary user cannot create API key for another principal.
+    """
+    with principals_context["context"] as context:
+        # Log in as Bob, this API endpoint is unauthorized
+        with enter_password("secret2"):
+            context.authenticate(username="bob")
+
+        principal_uuid = principals_context["uuid"][username]
+        with fail_with_status_code(401):
+            _create_api_key_other_principal(
+                context=context, uuid=principal_uuid, scopes=["read:metadata"]
+            )
+
+
+def test_api_key_bypass_scopes(enter_password, principals_context):
+    """
+    Ordinary user cannot create API key that bypasses a scopes restriction.
+    """
+    with principals_context["context"] as context:
+        # Log in as Bob, create API key with empty scopes
+        with enter_password("secret2"):
+            context.authenticate(username="bob")
+
+        response = context.http_client.post(
+            "/api/v1/auth/apikey", json={"expires_in": None, "scopes": []}
+        )
+        response.raise_for_status()
+        api_key = response.json()["secret"]
+        assert api_key
+        context.logout()
+
+        # Try the new API key with admin and normal resources
+        for resource in ("/api/v1/auth/principal", "/api/v1/array/full/A1"):
+            # Try with/without key, with/without empty scopes
+            for query_params in (
+                f"api_key={api_key}",
+                "scopes=[]",
+                f"api_key={api_key}&scopes=[]",
+            ):
+                url_query = f"{resource}?{query_params}"
+                with fail_with_status_code(401):
+                    context.http_client.get(url_query).raise_for_status()
```

Comment:
The `.get(...)` method takes a `params` dict, which you can use to avoid doing the string munging yourself here. Also, see comment above about preferring `context.api_key`.

Reply:
Nice tip!